gossip/src/crds_value.rs

```diff
@@ -163,6 +170,110 @@ impl CrdsValue {
+    ) -> Result<(), IoError>
+    where
+        I: IntoIterator<Item = T>,
+        <I as IntoIterator>::IntoIter: ExactSizeIterator,
+    {
+        let mut values = values.into_iter();
+        let size = u64::try_from(values.len()).unwrap();
+        writer.write_all(&size.to_le_bytes())?;
+        values.try_for_each(|value| value.borrow().bincode_serialize(writer))
+    }
+
+    // Implements bincode::deserialize_from for CrdsValue.
+    pub(crate) fn bincode_deserialize(
+        bytes: &[u8],
+        allow_trailing_bytes: bool,
+    ) -> Result<Self, bincode::Error> {
+        let (signature, bytes) = convert_fixed_bytes::<Signature, SIGNATURE_BYTES>(bytes)?;
+        let mut cursor = Cursor::new(bytes);
+        // Default bincode options:
+        //  * unlimited byte limit
+        //  * little endian
+        //  * varint encoding
+        //  * rejects trailing bytes
+        // https://docs.rs/bincode/1.3.3/bincode/fn.options.html
+        let options = bincode::options()
+            .with_limit(bytes.len() as u64)
+            .with_fixint_encoding();
+        let data: CrdsData = if allow_trailing_bytes {
+            options.allow_trailing_bytes().deserialize_from(&mut cursor)
+        } else {
+            options.deserialize_from(&mut cursor).and_then(|data| {
+                // We have to manually check if all bytes are read because
+                // options.reject_trailing_bytes() does not really work.
+                // https://github.com/bincode-org/bincode/issues/732
+                if cursor.position() == cursor.get_ref().len() as u64 {
+                    Ok(data)
+                } else {
+                    Err(bincode::Error::from(bincode::ErrorKind::Custom(
+                        String::from("Slice had bytes remaining after deserialization"),
+                    )))
+                }
+            })
+        }?;
+        let offset = usize::try_from(cursor.position()).map_err(|err| {
+            let err = format!("{err:?}, cursor: {}", cursor.position());
+            bincode::ErrorKind::Custom(err)
+        })?;
+        let bincode_serialized_data = bytes.get(..offset).map(Vec::from).ok_or_else(|| {
+            let err = format!("Invalid offset: {offset}, bytes.len(): {}", bytes.len());
+            bincode::ErrorKind::Custom(err)
+        })?;
+        let hash = solana_sdk::hash::hashv(&[signature.as_ref(), &bincode_serialized_data]);
+        Ok(Self {
+            signature,
+            data,
+            hash,
+            bincode_serialized_data,
+        })
+    }
+
+    // Implements bincode::deserialize_from for Vec<CrdsValue>.
+    // Verifies that there are exactly as many entries as the encoded size.
+    pub(crate) fn bincode_deserialize_many(
+        bytes: &[u8],
+    ) -> impl Iterator<Item = Result<Self, bincode::Error>> + '_ {
+        // Decode number of items in the slice.
+        let (size, mut bytes) = match convert_fixed_bytes::<[u8; 8], 8>(bytes) {
```

Comment:
`convert_fixed_bytes::<[u8; 8], 8>` what exactly is the last `8` doing here? can we be more descriptive here on what these `8`s represent?

Reply:
size is just the first 8 bytes of the serialized struct? and dictates how long the `Vec<CrdsValue>` is in terms of items (not bytes)?

Reply:
Added a comment to clarify.

```rust
// bincode::serialize{,_into} serializes sequence length as
// a u64 (8 bytes) with fixint encoding in little endian.
```